java-viewer/src/test/java/org/sonar/java/viewer/ViewerTest.java

```diff
@@ -77,4 +92,75 @@ public void values_with_error() {
     assertThat(values.get("errorStackTrace")).startsWith("java.lang.Exception: " + message);
   }
 
+  @Test
+  public void start_server_and_test_requests() throws Exception {
+    ServerSocket serverSocket = new ServerSocket(0);
+    int localPort = serverSocket.getLocalPort();
+    serverSocket.close();
+    String uri = "http://localhost:" + localPort + "/";
+    Viewer.startWebServer(localPort, "class A {void fun() {}}");
+    try(CloseableHttpClient client = HttpClients.createMinimal()) {
+      CloseableHttpResponse resp = client.execute(new HttpGet(uri));
+      assertThat(resp.getStatusLine().getStatusCode()).isEqualTo(200);
+      assertThat(EntityUtils.toString(resp.getEntity())).isEqualTo(IOUtils.toString(new FileInputStream(new File("src/test/resources/viewer_result1.html"))));
```

Comment:
Testing for equality of the whole input seems bit fragile to me. I think it will be enough to test that response contains the code which was provided. Also this means that any minor change to the html output requires changing 3 test resources.